### PR TITLE
Restclient

### PR DESCRIPTION
**Pull request by batshoes**
Initial step towards a BlockSpring integration. We'll iterate over this to make it more TDD

### CODE REVIEW

File: lib/engagements_count_per_site_or_operator.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    end_date = DateTime.parse(request.params['end_date']).iso8601
+    end_date = end_date.gsub "00:00:00", "23:59:59"
+
+    request_payload = prepare_payload(start_date, end_date, site_id, operator_id)
+    engagement_count_response = ApiRequest.new("#{token}").call(API_ENDPOINT, request_payload)
+    engagement_count = engagement_count_response.first['count']
+    response.addOutput("output", engagement_count)
+    response.end()
+  end
+Blockspring.define(block)
```
Comment (paniagua) [inline lib/engagements_count_per_site_or_operator.rb]: bad indentation

File: lib/operator_calls_per_site.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    end_date = DateTime.parse(request.params['end_date']).iso8601
+    end_date = end_date.gsub "00:00:00", "23:59:59"
+
+    request_payload =  prepare_payload(start_date, end_date, site_id, operator_id)
+    operator_calls_response = ApiRequest.new("#{token}").call(API_ENDPOINT, request_payload)
+    operator_calls_count = operator_calls_response.first['count']
+    response.addOutput("output", operator_calls_count)
+    response.end()
+  end
+Blockspring.define(block)
```
Comment (paniagua) [inline lib/operator_calls_per_site.rb]: same here

File: lib/total_missed_calls_per_site.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    request_payload = prepare_payload(start_date, end_date, site_id, "operator_accepted")
+    operator_accepted_count_response = ApiRequest.new("#{token}").call(API_ENDPOINT, request_payload)
+
+
+    total_missed_call_count = visitor_requested_count_response.first['count'] - operator_accepted_count_response.first['count']
+    response.addOutput("output", total_missed_call_count)
+    response.end()
+  end
+Blockspring.define(block)
```
Comment (paniagua) [inline lib/total_missed_calls_per_site.rb]: same

File: lib/visitors_count_per_site.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    end_date = DateTime.parse(request.params['end_date']).iso8601
+    end_date = end_date.gsub "00:00:00", "23:59:59"
+
+    request_payload = prepare_payload(start_date, end_date, site_id)
+    visitor_count_response = ApiRequest.new("#{token}").call(API_ENDPOINT, request_payload)
+    visitor_count = visitor_count_response.first['count']
+    response.addOutput("output", visitor_count)
+    response.end()
+  end
+Blockspring.define(block)
```
Comment (paniagua) [inline lib/visitors_count_per_site.rb]: one more

File: spec/blockspring_spec.rb [hunk 5]
```diff
@@ -9,3 +9,12 @@
     expect(false).to eq(true)
   end
 end
+
+describe Params
```
Comment (paniagua) [inline spec/blockspring_spec.rb]: does this file work? if it is a work in progress we can remove it from this PR

File: lib/engagements_count_per_site_or_operator.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
+require "blockspring"
+require 'rubygems' if RUBY_VERSION < '1.9'
```
Comment (paniagua) [inline lib/engagements_count_per_site_or_operator.rb]: why is the if necessary ?

File: lib/engagements_count_per_site_or_operator.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            type: "selector",
+            dimension: "operator_id",
+            value: "#{operator_id}"
+          }
+        ]
+      }
+    }
+  end
+
+  block  = lambda do |request, response|
```
Comment (paniagua) [inline lib/engagements_count_per_site_or_operator.rb]: two spaces between the ```block``` and the ```=```

File: lib/operator_calls_per_site.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            type: "selector",
+            dimension: "operator_id",
+            value: "#{operator_id}"
+          }
+        ]
+      }
+    }
+  end
+
+  block  = lambda do |request, response|
```
Comment (paniagua) [inline lib/operator_calls_per_site.rb]: two spaces

File: spec/blockspring_spec.rb (hunk 5 above)
Comment (batshoes) [inline spec/blockspring_spec.rb]: This was originally for the "Params" class that I made, I never removed the start of the test. It's been deleted.

File: lib/operator_online_duration.rb [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+            type: "selector",
+            dimension: "operator_id",
+            value: "#{operator_id}"
+          }
+        ]
+      }
+    }
+  end
+
+  block  = lambda do |request, response|
```
Comment (paniagua) [inline lib/operator_online_duration.rb]: 2 spaces

File: lib/total_missed_calls_per_site.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
+require "blockspring"
+require 'rubygems' if RUBY_VERSION < '1.9'
+require_relative 'api_call'
+
+module Blockspring
+
+  API_ENDPOINT = "https://api.salemove.com/visitors/stats/count"
+
+  def prepare_payload(start_date, end_date, site_id, type)
```
Comment (paniagua) [inline lib/total_missed_calls_per_site.rb]: is ```type``` a ```selector_type```  ? or what types are we talking about in here

File: lib/total_missed_calls_per_site.rb [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    end_date = end_date.gsub "00:00:00", "23:59:59"
+
+    request_payload = prepare_payload(start_date, end_date, site_id, "visitor_requested")
+    visitor_requested_count_response = ApiRequest.new("#{token}").call(API_ENDPOINT, request_payload)
+
+    request_payload = prepare_payload(start_date, end_date, site_id, "operator_accepted")
+    operator_accepted_count_response = ApiRequest.new("#{token}").call(API_ENDPOINT, request_payload)
+
+
+    total_missed_call_count = visitor_requested_count_response.first['count'] - operator_accepted_count_response.first['count']
```
Comment (paniagua) [inline lib/total_missed_calls_per_site.rb]: line too long, alternatively you can do 

```
requested_calls = visitor_requested_count_response.first['count']
accepted_calls = operator_accepted_count_response.first['count']
missed_call_count = requested_calls - accepted_calls
```

in general a line should be max 80 chars.

File: lib/engagements_count_per_site_or_operator.rb (hunk 1 above)
Comment (batshoes) [inline lib/engagements_count_per_site_or_operator.rb]: Not sure what you mean by bad indentation. Nothing here is really nested within something else. Bad spacing? 
```    site_id = request.params['site_id']
    operator_id = request.params['operator_id']
    token = request.params['token']

    start_date = DateTime.parse(request.params['start_date']).iso8601
    end_date = DateTime.parse(request.params['end_date']).iso8601
    end_date = end_date.gsub "00:00:00", "23:59:59"

    request_payload = prepare_payload(start_date, end_date, site_id, operator_id)
    engagement_count_response = ApiRequest.new("#{token}").call(API_ENDPOINT, request_payload)
    engagement_count = engagement_count_response.first['count']
    
    response.addOutput("output", engagement_count)
    response.end()```
Would something like this be better ?

File: lib/api_call.rb [hunk 12]
```diff
@@ -0,0 +1,18 @@
+require 'rest_client'
+require 'json'
+
+Class ApiRequest
```
Comment (indrekj) [inline lib/api_call.rb]: `Class` with a capital C?

File: lib/engagements_count_per_site_or_operator.rb [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+  block = lambda do |request, response|
+      
+    site_id = request.params['site_id']
+    operator_id = request.params['operator_id']
+    token = request.params['token']
+
+    start_date = DateTime.parse(request.params['start_date']).iso8601
+    end_date = DateTime.parse(request.params['end_date']).iso8601
+    end_date = end_date.gsub "00:00:00", "23:59:59"
```
Comment (indrekj) [inline lib/engagements_count_per_site_or_operator.rb]: Seems odd. end date is exclusive IIRC

Comment (batshoes) [inline lib/engagements_count_per_site_or_operator.rb]: Adjusted time to make it inclusive of the date entered. Realize now that I am coming into this with my own thoughts on a process that has already been decided. Will remove.

**Pull request merged**